src/client/MetaRequest.h

```diff
@@ -202,9 +202,15 @@ struct MetaRequest {
   bool auth_is_best() {
     if ((head.op & CEPH_MDS_OP_WRITE) || head.op == CEPH_MDS_OP_OPEN ||
         (head.op == CEPH_MDS_OP_GETATTR && (head.args.getattr.mask & CEPH_STAT_RSTAT)) ||
-	head.op == CEPH_MDS_OP_READDIR || send_to_auth) 
+	head.op == CEPH_MDS_OP_READDIR || head.op == CEPH_MDS_OP_MKNOD ||
+	head.op == CEPH_MDS_OP_SYMLINK || head.op == CEPH_MDS_OP_LINK ||
+	head.op == CEPH_MDS_OP_UNLINK || head.op == CEPH_MDS_OP_RENAME ||
+	head.op == CEPH_MDS_OP_SETLAYOUT || head.op == CEPH_MDS_OP_SETDIRLAYOUT ||
+	head.op == CEPH_MDS_OP_SETXATTR || head.op == CEPH_MDS_OP_RMXATTR ||
+	head.op == CEPH_MDS_OP_SETFILELOCK || head.op == CEPH_MDS_OP_SETATTR ||
```

Comment:
I think most of these seem to already be covered by the initial CEPH_MDS_OP_WRITE check. This looks like it's just adding dead code.

Reply:
> I think most of these seem to already be covered by the initial CEPH_MDS_OP_WRITE check. This looks like it's just adding dead code.

I didn't get this.

Such as for the `CEPH_MDS_OP_SETFILELOCK`, in the MDS side it will try to xlock the `flocklock` and at the same time it needs the auth pin, for both need to send requests to the auth MDS.

In `Client.cc` in `choose_target_mds()`, if the `CEPH_MDS_OP_SETFILELOCK` is not in the `auth_is_best()`, it will choose a random MDS.

This PR will sync to kclient too.

Did I miss something ?

Reply:
And by checking the kclient/libcephfs history commits, someone had hit and seen there were many request traffics between MDSes due to this issue.

Reply:
Note that the first test in this if statement is:

`(head.op & CEPH_MDS_OP_WRITE)`

...and CEPH_MDS_OP_WRITE is defined as:

`#define CEPH_MDS_OP_WRITE        0x001000`

...so:

`CEPH_MDS_OP_SETFILELOCK=0x01109,`

...so that op is already covered by the check for OP_WRITE. AFAICT, all of the extra conditions you're adding will never be reached due to the OP_WRITE check.

Reply:
Okay, you are right. I didn't notice it.